crates/shared/src/price_estimation/trade_verifier.rs

```diff
@@ -508,25 +662,51 @@ impl SettleOutput {
 fn ensure_quote_accuracy(
     inaccuracy_limit: &BigRational,
     query: &PriceQuery,
-    solver: H160,
+    trade: &TradeKind,
     summary: &SettleOutput,
-) -> Result<Estimate, Error> {
+) -> std::result::Result<Estimate, Error> {
     // amounts verified by the simulation
     let (sell_amount, buy_amount) = match query.kind {
         OrderKind::Buy => (summary.out_amount, query.in_amount.get()),
         OrderKind::Sell => (query.in_amount.get(), summary.out_amount),
     };
+    let (sell_amount, buy_amount) = (
+        u256_to_big_rational(&sell_amount),
+        u256_to_big_rational(&buy_amount),
+    );
+
+    // Calculate the expected token losses that should be covered by JIT orders.
+    let (expected_sell_token_lost, expected_buy_token_lost) = match trade {
+        TradeKind::Regular(trade) if !trade.jit_orders.is_empty() => {
```

Comment:
While it should be possible to compute these values without changing the simulation logic maybe it's simpler to alter the simulation logic instead.
You could pass the `tokens` vector into the `Solver.swap()` function and it could store the settlement contract balance of each token before and after the settlement. This would results in a balances vector like: `[...settlement_contract_balances_before, user_out_token_balance_before, user_out_token_balance_after, ...settlement_contract_balances_after]`.

Then in `SettleOutput::decode()` you could compute the settlement balance difference for each token in addition to the `out_amount` of the user.

Fetching the trader balances would have to be tweaked very slightly. Instead of fetching the balances as the first and last middle interaction. They would have to be fetched as the last pre-interaction and the first post-interaction. That way we would be able to see the impact of fee-on-transfer tokens but don't get accounting errors if the user receives additional in a pre-interaction.

Would make sense to address this comment before adjusting the tests because the tested interfaces should simplify significantly.

Reply:
Updated the code with only one caveat:
> Fetching the trader balances would have to be tweaked very slightly. Instead of fetching the balances as the first and last middle interaction. They would have to be fetched as the last pre-interaction and the first post-interaction.

When doing so, the buy token balance shows `0`. So I had to revert this change: [`bea5183` (#3085)](https://github.com/cowprotocol/services/pull/3085/commits/bea518322ed073ac3ea8afa07f4b382a105473f4)

Reply:
Looks like the logic to decide which address to check the token balance we should look up for needs to be changed as well for this to work. Previously we would check the receiver or settlement contract balance but with that change we should check the balance for the receiver or `from` address. All the settlement contract balances are covered by the balance queries for all tokens in the price vector.

Reply:
Makes sense. That helped, thanks.

Reply:
Makes sense. That helped, thanks.